nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java

```diff
@@ -0,0 +1,290 @@
+     *
+     * @param schemaDefinition       the schema definition to validate
+     * @param isRootSchemaDefinition set to if schema definition is a root definition
+     * @throws IllegalArgumentException if any identifier does not end with .proto extension
+     */
+    public void validateSchemaDefinitionIdentifiers(final SchemaDefinition schemaDefinition, final boolean isRootSchemaDefinition) {
+        // do not validate schema identifier names for root schema definitions. They might be coming from sources like text fields,
+        // flow file attributes and other sources that do not support naming.
+        if (!isRootSchemaDefinition) {
+            validateSchemaIdentifier(schemaDefinition.getIdentifier());
+        }
+
+        // Recursively validate all referenced schemas
+        // schema references have to end with .proto extension.
+        for (final SchemaDefinition referencedSchema : schemaDefinition.getReferences().values()) {
+            validateSchemaDefinitionIdentifiers(referencedSchema, false);
+        }
+    }
+
+    /**
+     * Validates that a single SchemaIdentifier has a name ending with .proto extension.
+     *
+     * @param schemaIdentifier the schema identifier to validate
+     * @throws IllegalArgumentException if the identifier name does not end with .proto extension
+     */
+    private void validateSchemaIdentifier(final SchemaIdentifier schemaIdentifier) {
+        final Optional<String> nameOptional = schemaIdentifier.getName();
+        if (nameOptional.isPresent()) {
+            final String name = nameOptional.get();
+            if (!name.endsWith(".proto")) {
+                throw new IllegalArgumentException(
+                    String.format("Schema identifier name '%s' must end with .proto extension", name));
+            }
+        } else {
+            throw new IllegalArgumentException("Schema identifier must have a name that ends with .proto extension");
+        }
+    }
+
+    /**
+     * Writes a schema definition to the temporary directory structure.
+     * If package name is present, creates the appropriate directory structure.
+     *
+     * @param tempDir          the temporary directory root
+     * @param schemaDefinition the schema definition to write
+     * @throws IOException if unable to create directories or write files
+     */
+    private void writeSchemaToTempDirectory(final Path tempDir, final SchemaDefinition schemaDefinition) throws IOException {
+        logger.debug("Writing schema definition to temporary directory. Identifier: {}", schemaDefinition.getIdentifier());
+
+        final String schemaFileName = generateSchemaFileName(schemaDefinition);
+        final Path schemaFile = tempDir.resolve(schemaFileName);
+
+        // Write schema text to file
+        Files.write(schemaFile, schemaDefinition.getText().getBytes(), StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+        logger.debug("Successfully wrote schema to file: {} (size: {} bytes)",
```

Comment:
AFAIK there is a subtle difference between bytes length and `String.length()`
```suggestion
        logger.debug("Successfully wrote schema to file: {} (string length: {})",
```